### PR TITLE
Framework for creating and using the Unity Catalog connections API

### PR DESCRIPTION
**Pull request by jamesd-db**
Creating a framework for creating and using the Unity Catalog connections API with the Databricks CLI.

### CODE REVIEW

File: databricks_cli/unity_catalog/connection_cli.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+from databricks_cli.click_types import JsonClickType
+from databricks_cli.configure.config import provide_api_client, profile_option, debug_option
+from databricks_cli.unity_catalog.api import UnityCatalogApi
+from databricks_cli.unity_catalog.utils import del_none, hide, json_file_help, json_string_help, \
+    mc_pretty_format
+from databricks_cli.utils import eat_exceptions, CONTEXT_SETTINGS, json_cli_base
+
+# These two options are shared among create and updates, so they are very common
+def create_update_common_options(f):
+    @click.option('--read-only/--no-read-only', is_flag=True, default=None,
```
Comment (andrewli81) [inline databricks_cli/unity_catalog/connection_cli.py:37]: nit: 'Whether the connection is read-only'

File: databricks_cli/unity_catalog/connection_cli.py [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='Create online catalog connection with CLI flags.')
+@common_create_args
+@create_update_common_options
+@debug_option
+@eat_exceptions
+@profile_option
+@provide_api_client
+def create_online_catalog_cli(api_client, name, host, port, user,
```
Comment (andrewli81) [inline databricks_cli/unity_catalog/connection_cli.py:307]: Ah I forgot to say this is not user creatable!

File: databricks_cli/unity_catalog/connection_cli.py [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+@debug_option
+@eat_exceptions
+@profile_option
+@provide_api_client
+def create_json(api_client, json_file, json):
+    '''
+    Create new connection with an inline JSON or JSON file input.
+    '''
+    if (json is None) and (json_file is None):
+        raise ValueError('Must either provide inline JSON or JSON file.')
```
Comment (andrewli81) [inline databricks_cli/unity_catalog/connection_cli.py:337]: nit: "Must either provide inline JSON or JSON file *path*."

File: databricks_cli/unity_catalog/connection_cli.py [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+        
+
+
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='List connections.')
+@debug_option
+@profile_option
+@eat_exceptions
+@provide_api_client
+def list_connections_cli(api_client, ):
```
Comment (andrewli81) [inline databricks_cli/unity_catalog/connection_cli.py:351]: you don't need the extra comma?

File: databricks_cli/unity_catalog/connection_cli.py [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+@click.command(context_settings=CONTEXT_SETTINGS,
+               short_help='Delete a connection.')
+@click.option('--name', required=True,
+              help='Name of the connection to delete.')
+@debug_option
+@profile_option
+@eat_exceptions
+@provide_api_client
+def delete_connection_cli(api_client, name):
```
Comment (andrewli81) [inline databricks_cli/unity_catalog/connection_cli.py:383]: Did we omit the update_connection_cli?

File: tests/unity_catalog/test_con_cli.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+from click.testing import CliRunner
+
+from databricks_cli.unity_catalog import connection_cli
+from tests.utils import provide_conf
+
+CONNECTION_NAME = 'test_connection_name'
+COMMENT = 'some_comment'
+
+TESTHOST = "test_postgresql.fakedb.com"
+TESTHOST2 = "postgresql.fakedb2.lan"
```
Comment (andrewli81) [inline tests/unity_catalog/test_con_cli.py:38]: TESTHOST2 is unused?

File: tests/unity_catalog/test_con_cli.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+
+@pytest.fixture()
+def echo_mock():
+    with mock.patch('databricks_cli.unity_catalog.connection_cli.click.echo') as echo_mock:
+        yield echo_mock
+
+
+@provide_conf
+def test_create_connection_cli(api_mock):
```
Comment (andrewli81) [inline tests/unity_catalog/test_con_cli.py:87]: Can we add tests for get, update, delete list as well?

**Comment by andrewli81** [comment]
@jamesd-db Good news - @pietern is working on a new Databricks CLI that automatically generates client code based on open api spec - so we will have coverage automatically. Looks like it's already showing on the SDK: https://github.com/databricks/databricks-sdk-go/blob/1504930d8c20337f353f13d4934700d85e0b6ed5/service/catalog/interface.go#L177-L220

We can stop polishing this PR or adding tests and just use this as a test environment for our OAuth work!

**Comment by alexott** [comment]
I think that it's better to switch to the new [Databricks CLI](https://github.com/databricks/cli) and Python SDKs that are automatically generated from OpenAPI specs

**Comment by andrewli81** [comment]
@alexott we chatted with @pietern - we just want to archive this code (it is mainly used for testing oauth support for query federation connections) in its dedicated branch since it was worked on by our intern before the new cli was live and also we might not have time to convert it into the new cli during the internship. We eventually will move this to the new cli but likely after @jamesd-db's internship ends. So @pietern created this branch for us.

**Comment by pietern** [comment]
By merging we archive the work on a branch for future reference. No intention of merging this into main.

**Comment by andrewli81** [comment]
@pietern Yes this PR is otherwise done, @jamesd-db uploaded the OAuth client changes and pointed this PR to the branch you created, we should be good to merge.

**Comment by jamesd-db** [comment]
jenkins merge

**Comment by pietern** [comment]
References to JIRA: SC-133066 and SC-134677

**Pull request merged**